In [20]:
import numpy as np
import pandas as pd

In [21]:
reldir = 'C:\\Users\\shefa\\OneDrive\\CS6220\\msd.recsys\\datasets\\'

In [22]:
standardized_df = pd.read_pickle(reldir + 'combined.csv', compression='zip')

In [23]:
master_df = pd.read_pickle(reldir + 'songdata.csv', compression='zip')

In [24]:
standardized_df.head()

user                song  count  \
0  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOWEZSI12A81C21CE6      1   
1  833c530ecda3d99deb8395f70400aa3999783d91  SOWEZSI12A81C21CE6      2   
2  d6c5bd2b570b4faf8964d7ed04f3392ff505d2be  SOWEZSI12A81C21CE6      1   
3  724534729c9f5dc72a009269c2c225883e4775d2  SOWEZSI12A81C21CE6      1   
4  ee7aa84c164038c963cfd02a7e52a5598aa470c3  SOWEZSI12A81C21CE6      2   

  analysis_songs_track_id metadata_songs_artist_id metadata_songs_song_id  \
0      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
1      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
2      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
3      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
4      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   

   analysis_songs_key  analysis_songs_key_confidence  analysis_songs_mode  \
0                   5                          0.351                    0   
1                   5                          0.351                    0   
2                   5                          0.351                    0   
3                   5                          0.351                    0   
4                   5                          0.351                    0   

   analysis_songs_mode_confidence  ...  tempo_binned  analysis_songs_duration  \
0                           0.318  ...             8                -0.280898   
1                           0.318  ...             8                -0.280898   
2                           0.318  ...             8                -0.280898   
3                           0.318  ...             8                -0.280898   
4                           0.318  ...             8                -0.280898   

   metadata_songs_artist_familiarity  metadata_songs_artist_hotttnesss  \
0                           0.732042                          1.243373   
1                           0.732042                          1.243373   
2                           0.732042                          1.243373   
3                           0.732042                          1.243373   
4                           0.732042                          1.243373   

   metadata_songs_song_hotttnesss  \
0                         5.09644   
1                         5.09644   
2                         5.09644   
3                         5.09644   
4                         5.09644   

                               metadata_artist_terms  \
0  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
1  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
2  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
3  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
4  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   

   metadata_songs_artist_name  metadata_songs_release  metadata_songs_title  \
0                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
1                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
2                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
3                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
4                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   

         musicbrainz_artist_mbtags  
0  b'classic pop and rock'|b'folk'  
1  b'classic pop and rock'|b'folk'  
2  b'classic pop and rock'|b'folk'  
3  b'classic pop and rock'|b'folk'  
4  b'classic pop and rock'|b'folk'  

[5 rows x 27 columns]

https://jessesw.com/Rec-System/

https://towardsdatascience.com/building-a-collaborative-filtering-recommender-system-with-clickstream-data-dffc86c8c65

In [31]:
master_df = standardized_df[standardized_df.groupby('user').user.transform('count')>4].copy()
master_df.head()

user                song  count  \
3   724534729c9f5dc72a009269c2c225883e4775d2  SOWEZSI12A81C21CE6      1   
4   ee7aa84c164038c963cfd02a7e52a5598aa470c3  SOWEZSI12A81C21CE6      2   
5   1aa4fd215aadb160965110ed8a829745cde319eb  SOWEZSI12A81C21CE6      1   
8   cca2694a19b2474bc731574ebb8b10382214f865  SOWEZSI12A81C21CE6      2   
17  091d959cd1d21f270d9f6dc58b93ee78caa4ab96  SOWEZSI12A81C21CE6     10   

   analysis_songs_track_id metadata_songs_artist_id metadata_songs_song_id  \
3       TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
4       TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
5       TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
8       TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
17      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   

    analysis_songs_key  analysis_songs_key_confidence  analysis_songs_mode  \
3                    5                          0.351                    0   
4                    5                          0.351                    0   
5                    5                          0.351                    0   
8                    5                          0.351                    0   
17                   5                          0.351                    0   

    analysis_songs_mode_confidence  ...  metadata_songs_artist_familiarity  \
3                            0.318  ...                           0.732042   
4                            0.318  ...                           0.732042   
5                            0.318  ...                           0.732042   
8                            0.318  ...                           0.732042   
17                           0.318  ...                           0.732042   

    metadata_songs_artist_hotttnesss  metadata_songs_song_hotttnesss  \
3                           1.243373                         5.09644   
4                           1.243373                         5.09644   
5                           1.243373                         5.09644   
8                           1.243373                         5.09644   
17                          1.243373                         5.09644   

                                metadata_artist_terms  \
3   b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
4   b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
5   b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
8   b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
17  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   

    metadata_songs_artist_name  metadata_songs_release  metadata_songs_title  \
3                  Gipsy Kings           Greatest Hits     Tu Quieres Volver   
4                  Gipsy Kings           Greatest Hits     Tu Quieres Volver   
5                  Gipsy Kings           Greatest Hits     Tu Quieres Volver   
8                  Gipsy Kings           Greatest Hits     Tu Quieres Volver   
17                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   

          musicbrainz_artist_mbtags  user_id  song_id  
3   b'classic pop and rock'|b'folk'   186356     3214  
4   b'classic pop and rock'|b'folk'   389460     3214  
5   b'classic pop and rock'|b'folk'    43196     3214  
8   b'classic pop and rock'|b'folk'   334055     3214  
17  b'classic pop and rock'|b'folk'    14761     3214  

[5 rows x 29 columns]

## Attempt 1 Using jessesw.com

In [6]:
## Check that all counts in dataset are 1 or greater
standardized_df.describe()

count  analysis_songs_key  analysis_songs_key_confidence  \
count  772661.000000       772661.000000                  772661.000000   
mean        2.684340            5.673522                       0.432972   
std         5.454646            3.615617                       0.270784   
min         1.000000            0.000000                       0.000000   
25%         1.000000            2.000000                       0.222000   
50%         1.000000            6.000000                       0.437000   
75%         2.000000            9.000000                       0.630000   
max       771.000000           11.000000                       1.000000   

       analysis_songs_mode  analysis_songs_mode_confidence  \
count        772661.000000                   772661.000000   
mean              0.637005                        0.489415   
std               0.480864                        0.165614   
min               0.000000                        0.000000   
25%               0.000000                        0.397000   
50%               1.000000                        0.480000   
75%               1.000000                        0.596000   
max               1.000000                        1.000000   

       analysis_songs_tempo  analysis_songs_time_signature  \
count         772661.000000                  772661.000000   
mean             126.560860                       3.730847   
std               31.789887                       1.021593   
min                0.000000                       0.000000   
25%              102.984000                       4.000000   
50%              121.884000                       4.000000   
75%              146.022000                       4.000000   
max              253.357000                       7.000000   

       analysis_songs_time_signature_confidence  musicbrainz_songs_year  \
count                             772661.000000           772661.000000   
mean                                   0.626652             1543.714245   
std                                    0.361508              839.540567   
min                                    0.000000                0.000000   
25%                                    0.358000             1977.000000   
50%                                    0.742000             2000.000000   
75%                                    0.994000             2006.000000   
max                                    1.000000             2010.000000   

        years_binned   tempo_binned  analysis_songs_duration  \
count  772661.000000  772661.000000            772661.000000   
mean        2.646629       4.783883                 0.158297   
std         1.620828       2.637524                 0.801748   
min         0.000000       0.000000                -2.151904   
25%         1.000000       3.000000                -0.228571   
50%         3.000000       5.000000                 0.052587   
75%         4.000000       7.000000                 0.464172   
max         4.000000       9.000000                14.586918   

       metadata_songs_artist_familiarity  metadata_songs_artist_hotttnesss  \
count                      772661.000000                     772661.000000   
mean                            1.137810                          1.828713   
std                             0.594691                          1.180343   
min                            -2.813196                         -2.960846   
25%                             0.847774                          1.077402   
50%                             1.298247                          1.707256   
75%                             1.572213                          2.589013   
max                             2.176964                          5.457244   

       metadata_songs_song_hotttnesss  
count                   772661.000000  
mean                         3.760755  
std                          2.317352  
min                         -4.007280  
25%                          2.414042  
50%                      

In [7]:
## Get our unique users
users = list(np.sort(standardized_df.user.unique()))
## Get our unique songs
songs = list(standardized_df.song.unique())
## Get counts of songs listened to
song_count = list(standardized_df['count'])

In [8]:
import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve

## Get associated row indices
rows = pd.Categorical(standardized_df.user, categories=users).codes
## Get associated column indices
cols = pd.Categorical(standardized_df.song, categories=songs).codes
matrix = sparse.csr_matrix((song_count, (rows, cols)), shape=(len(users), len(songs)))

In [9]:
matrix

<418252x3675 sparse matrix of type '<class 'numpy.intc'>'
	with 772661 stored elements in Compressed Sparse Row format>

We have a total of 418,252 users and 3,675 songs, with 772,661 songs heard.

In [10]:
## Calculate matrix sparsity
# Obtain number of possible interactions in the matrix
matrix_size = matrix.shape[0]*matrix.shape[1]
# Obtain number of songs interacted with
num_songs_heard = len(matrix.nonzero()[0])
sparsity = 100*(1-(num_songs_heard/matrix_size))
sparsity

99.94973176669653

99.9% of the interaction matrix is sparse whereas the maximum that usually works for collaborative filtering is 99.5%. Therefore, we can attempt to build a matrix factorization model, but this may not perform well.

In [11]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install implicit

## Attempt 2 Using Towards Data Science

In [63]:
#content_matrix = sparse.csr_matrix((song_count, (rows, cols)), shape=(len(songs), len(users)))
import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve
from scipy.sparse import csr_matrix

master_df['user']= master_df['user'].astype("category")
master_df['song']= master_df['song'].astype("category")
master_df['user_id']= master_df['user'].cat.codes
master_df['song_id']= master_df['song'].cat.codes

sparse_content_user = sparse.csr_matrix((master_df['count'].astype(float), (master_df['song_id'], master_df['user_id'])))
sparse_user_content = sparse.csr_matrix((master_df['count'].astype(float), (master_df['user_id'], master_df['song_id'])))
print (sparse_user_content)

  (6, 639)	2.0
  (6, 1223)	1.0
  (6, 2370)	1.0
  (6, 2550)	1.0
  (6, 3548)	1.0
  (12, 114)	1.0
  (12, 370)	5.0
  (12, 868)	1.0
  (12, 1036)	7.0
  (12, 1192)	1.0
  (12, 2529)	5.0
  (12, 3374)	1.0
  (14, 1054)	1.0
  (14, 1231)	4.0
  (14, 1424)	2.0
  (14, 2276)	5.0
  (14, 2582)	1.0
  (14, 2792)	2.0
  (33, 114)	1.0
  (33, 964)	2.0
  (33, 1424)	9.0
  (33, 2792)	1.0
  (33, 2880)	8.0
  (33, 3213)	1.0
  (33, 3475)	1.0
  :	:
  (418219, 1831)	7.0
  (418219, 1965)	1.0
  (418219, 2142)	1.0
  (418219, 2165)	1.0
  (418219, 2374)	1.0
  (418219, 2426)	1.0
  (418219, 3311)	1.0
  (418219, 3634)	2.0
  (418226, 692)	1.0
  (418226, 924)	1.0
  (418226, 964)	1.0
  (418226, 1831)	1.0
  (418226, 2214)	1.0
  (418227, 114)	1.0
  (418227, 1153)	1.0
  (418227, 1215)	7.0
  (418227, 1235)	1.0
  (418227, 2545)	1.0
  (418227, 3467)	2.0
  (418227, 3542)	1.0
  (418246, 521)	4.0
  (418246, 1918)	4.0
  (418246, 2426)	13.0
  (418246, 3223)	14.0
  (418246, 3258)	19.0


In [47]:
## Create a dataframe to look up user
user_lookup = master_df[['user', 'user_id']].drop_duplicates()
user_lookup.head()

user  user_id
3   724534729c9f5dc72a009269c2c225883e4775d2   186356
4   ee7aa84c164038c963cfd02a7e52a5598aa470c3   389460
5   1aa4fd215aadb160965110ed8a829745cde319eb    43196
8   cca2694a19b2474bc731574ebb8b10382214f865   334055
17  091d959cd1d21f270d9f6dc58b93ee78caa4ab96    14761

In [48]:
## Create a dataframe to look up song
song_lookup = master_df[['song', 'song_id']].drop_duplicates()
song_lookup.head()

song  song_id
3     SOWEZSI12A81C21CE6     3214
771   SODCXXY12AB0187452      521
1364  SOWPAXV12A67ADA046     3269
5056  SOLXDDC12A6701FBFD     1853
7423  SONJBQX12A6D4F8382     2047

In [49]:
## Subset to data for model
model_df = master_df[['user_id', 'song_id', 'count']]
model_df.head()

user_id  song_id  count
3    186356     3214      1
4    389460     3214      2
5     43196     3214      1
8    334055     3214      2
17    14761     3214     10

In [ ]:
import implicit
## will need to tweak the number of latent factors by potentially identifying loss function stability
## will need to tweak regularization to tune parameters
## user cross validation to tune the optimum factors in model
## Reasonable solution usually occurs in 20 iterations or less
## Knowing the number of latent or hidden factors influencing the model is impossible. We should
## start at 5-10, then increase by 5 until performance of model stops improving
from sklearn.model_selection import StratifiedKFold
k_latent_factors = [5, 6, 7, 8, 9, 10, 15, 20, 25, 30, 35, 40]
regularization_options = [0.02, 0.05, 0.08, 0.1, 0.2, 0.3, 0.4]
k_iterations = [3, 6, 9, 12, 15, 18, 20]
alpha_options = [30, 35, 40, 45, 50]
col_names=['Model', 'Hit Ratio', 'Mean_Avg_Precision', 'Mean_Avg_Recall']
x = model_df[['song_id', 'count']].to_numpy()
y = model_df[['user_id']].to_numpy()
model_evaluation_metrics_df = pd.DataFrame(columns=col_names)
n_splits = 5
skf = KFold(n_splits=n_splits, shuffle=True, random_state=None)
skf.get_n_splits(model_df)
MAP = 0
for k in k_latent_factors:
    for i in regularization_options:
        for m in alpha_options:
            for train_index, test_index in skf.split(x, y):
                x_train, x_test = x[train_index], x[test_index]
                y_train, y_test = y[train_index], y[test_index]
                song_info = x_train[:,0]
                count_info = x_train[:,1]
                sparse_content_user = sparse.csr_matrix((count_info.astype(float), (song_info, y_train.flatten())))
                sparse_user_content = sparse.csr_matrix((count_info.astype(float),(y_train.flatten(), song_info)))
                model = implicit.als.AlternatingLeastSquares(factors=k, regularization=i, iterations=30)
                model.show_progress = False
                alpha = m
                data = (sparse_content_user*alpha).astype('double')
                model.fit(data)
                model_results = pd.DataFrame(data=x_test, columns=["song_id", 'count'])
                model_results['user_id']=pd.Series(y_test.flatten())
                model_results['predicted'] = model_results.apply(lambda row: model.recommend(row.user_id, sparse_user_content, N=10), axis=1)
                model_results['hits'] = model_results.apply(lambda row:[i for i in row['predicted'] if row['song_id'] in i], axis=1)
                model_results['hit'] = model_results.apply(lambda row: len(row['hits'])>0, axis=1)
                hits = len(model_results[model_results.hit==True])
                total = len(model_results)
                hit_ratio = hits/total
                user_actions = model_results.groupby('user_id').count()
                accurate_recommendations = model_results.groupby('user_id').sum()
                user_actions=user_actions[['song_id']]
                accurate_recommendations = accurate_recommendations[['hit']]
                metrics = user_actions.merge(accurate_recommendations, left_index=True, right_index=True)
                metrics['precision'] = metrics.hit/metrics.song_id
                metrics['recall'] = metrics.hit/10
                mean_average_precision = metrics.precision.mean()
                mean_average_recall = metrics.recall.mean()
                model_name = "factors:" + str(k) + " l2: " + str(i) + " alpha" + str(m)
                instance = {'Model': model_name, 'Hit Ratio': hit_ratio, 'Mean_Avg_Precision': mean_average_precision, 'Mean_Avg_Recall': mean_average_recall}
                model_evaluation_metrics_df = model_evaluation_metrics_df.append(instance, ignore_index=True)

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

Model  Hit Ratio  Mean_Avg_Precision  Mean_Avg_Recall
0  factors:10 l2: 0.02 alpha40  0.285185   0.292011            0.050557       
1  factors:10 l2: 0.02 alpha40  0.288462   0.297038            0.050893       
2  factors:10 l2: 0.02 alpha40  0.292973   0.299281            0.051615       
3  factors:10 l2: 0.02 alpha40  0.288405   0.294355            0.051154       
4  factors:10 l2: 0.02 alpha40  0.290656   0.296247            0.051435

Link to help explain confidence and alpha's impact on confidence:
https://stats.stackexchange.com/questions/155296/alternating-least-squares-als-why-two-different-kinds-of-setting-for-lambd

In [29]:
## The paper used to develop ALS used alpha of 40
alpha = 40
data = (sparse_content_user*alpha).astype('double')
model.fit(data)

  0%|          | 0/20 [00:00<?, ?it/s]

In [30]:
## Obtain recommendations for a given user
user_id = 350627
recommended = model.recommend(user_id, sparse_user_content, N=5)
print(recommended)

[(693, 0.79169756), (1722, 0.71202666), (203, 0.70730686), (2730, 0.6374319), (2942, 0.63132167)]


In [31]:
## Get similar items

song_id=2918
n_similar = 3
similar = model.similar_items(song_id, n_similar)
print(similar)

[(2918, 0.99999994), (162, 0.8860025), (227, 0.8719206)]


https://campus.datacamp.com/courses/recommendation-engines-in-pyspark/what-if-you-dont-have-customer-ratings?ex=6

In [ ]:
#Use ROEM as evaluation metric